app/src/main/java/org/mozilla/rocket/home/pinsite/PinSiteManager.kt

```diff
@@ -0,0 +1,277 @@
+    }
+
+    private fun load(results: MutableList<Site>) {
+        if (!this.isEnabled) {
+            log("load - not enabled")
+            return
+        }
+
+        log("load - enabled")
+        results.clear()
+
+        val isFirstInit = isFirstInit()
+        if (isFirstInit && partnerList.isNotEmpty()) {
+            results.addAll(0, partnerList)
+            log("load partner list")
+            save(results)
+        } else {
+            log("load saved pin site pref")
+            loadSavedPinnedSite(results)
+        }
+
+        if (isFirstInit) {
+            log("init finished")
+            onFirstInitComplete()
+        }
+    }
+
+    private fun initForUpdateUser(results: MutableList<Site>, partnerSites: List<Site>) {
+        results.addAll(partnerSites)
+    }
+
+    private fun initForNewUser(results: MutableList<Site>, partnerSites: List<Site>) {
+        results.addAll(partnerSites)
+
+        val defaultTopSiteJson = TopSitesUtils.loadDefaultSitesFromAssets(context, R.raw.topsites)
+        val defaultTopSites = jsonToSites(JSONArray(defaultTopSiteJson), true).toMutableList()
+
+        var remainPinCount = DEFAULT_NEW_USER_PIN_COUNT - partnerSites.size
+        while (remainPinCount-- > 0 && defaultTopSites.isNotEmpty()) {
+            results.add(defaultTopSites.removeAt(0))
+        }
+
+//        save(results)
+//        onFirstInitComplete()
+    }
+
+    private fun loadSavedPinnedSite(results: MutableList<Site>) {
+        pref.getString(KEY_STRING_JSON, "")?.let {
+            results.addAll(jsonToSites(JSONArray(it), false))
+        }
+    }
+
+    private fun isFirstInit(): Boolean {
+        return pref.getBoolean(KEY_BOOLEAN_FIRST_INIT, true)
+    }
+
+    private fun onFirstInitComplete() {
+        pref.edit().putBoolean(KEY_BOOLEAN_FIRST_INIT, false).apply()
+    }
+
+    private fun sitesToJson(sites: List<Site>): JSONArray {
+        val array = JSONArray()
+        for (i in sites.indices) {
+            val site = sites[i]
+            val jsonSite = siteToJson(site)
+            if (jsonSite != null) {
+                array.put(jsonSite)
+            }
+        }
+        return array
+    }
+
+    private fun jsonToSites(array: JSONArray, isDefaultTopSite: Boolean): List<Site> {
+        val sites = ArrayList<Site>()
+        val faviconPrefix = if (isDefaultTopSite) {
+            TopSitesUtils.TOP_SITE_ASSET_PREFIX
+        } else {
+            ""
+        }
+        try {
+            for (i in 0 until array.length()) {
+                val obj = array.getJSONObject(i)
+                sites.add(Site(obj.getLong(TopSitesUtils.KEY_ID),
+                        obj.getString(TopSitesUtils.KEY_TITLE),
+                        obj.getString(TopSitesUtils.KEY_URL),
+                        obj.getLong(TopSitesUtils.KEY_VIEW_COUNT),
+                        0,
+                        faviconPrefix + getFaviconUrl(obj)))
+            }
+        } catch (ignored: JSONException) {
```

Comment:
this will return an empty list of site. Maybe you'll just want to ignore the failing one.

Reply:
The first four parameters are essential for displaying that site, so any failure will make the item being skipped, while the last parameter can be empty, so getFaviconUrl will catch the exception itself and return an empty string.